File: Array/3sum.py
from typing import List
import copy

# 520ms 95.99%
# 18MB 7.87%
# 可以使用set类  add函数添加
class Solution:


    def threeSum(self, nums: List[int]) -> List[List[int]]:
        nums.sort()
        ans = []
        for i in range(len(nums)-2):
            target = -nums[i]
            # 剪枝1
            if i > 0 and nums[i-1] == nums[i]:
                continue

            # 剪枝2
            if nums[i+1] + nums[i+1] > target or nums[-1] + nums[-2] < target:
                continue

            left, right = i + 1, len(nums)-1
            while left < right:
                if nums[left] + nums[right] < target:
                    left += 1
                elif nums[left] + nums[right] > target:
                    right -= 1
                else:
                    # 去重
                    if len(ans) == 0 or ans[-1] != [-target, nums[left], nums[right]]:
                        ans.append([-target, nums[left], nums[right]])
                    left += 1
                    right -= 1
        return ans

if __name__ == '__main__':

    l = [-2, -1, 0, 0, 2, 2]
    a = Solution()
    print(a.threeSum(l))

File: Array/3sum.cpp
#include <iostream>
#include <algorithm>
#include <vector>
#include <map>
#include <set>
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        sort(nums.begin(), nums.end());
        set<vector<int>> ans;
        vector<vector<int>> ans_vec;
        int anchor = 0;
        while (anchor < nums.size() - 2){
            int left = anchor + 1, right = nums.size()-1;
            while (left < right){
                if (nums[anchor] + nums[left] + nums[right] == 0){
                    ans.insert({nums[anchor], nums[left], nums[right]});
                    left += 1;
                }
                else if (nums[anchor] + nums[left] + nums[right] < 0){
                    left += 1;
                } else{
                    right -= 1;
                }
            }
            anchor += 1;
        }
        for (auto it = ans.begin(); it != ans.end(); it++){
            ans_vec.push_back(*it);
        }
        return ans_vec;
    }
};

int main() {
    Solution sol;
    vector<int> nums = {-4, -1, -1, -1, 0 , 1, 2, 3};
    vector<vector<int>> ans = sol.threeSum(nums);
    for (int i = 0; i < ans.size(); i++){
        for (int j = 0; j < ans[i].size(); j++){
            cout << ans[i][j] << " ";
        }
        cout << endl;
    }
    return 0;
}